std/range/package.d

```diff
@@ -7232,6 +7245,1057 @@ if (isForwardRange!Source && hasLength!Source)
+            }
+        }
+        else static if (hasLength!Source)
+        {
+            auto opIndex(size_t index)
+            {
+                import std.algorithm.comparison : min;
+
+                immutable start = index * _stepSize;
+                immutable end = start + _windowSize;
+                immutable len = _source.length;
+                assert(start < len, "slides index out of bounds");
+                return _source[start .. min(end, len)];
+            }
+        }
+        // hasSlicing implies either isInfinite or hasLength
+        //else static if (hasSliceToEnd)
+        //{
+            //auto opIndex(size_t index)
+            //{
+                //return _source[index .. $].take(windowSize);
+            //}
+        //}
+
+        /// Ditto
+        static if (hasLength!Source)
+        {
+            typeof(this) opSlice(size_t lower, size_t upper)
+            {
+                import std.algorithm.comparison : min;
+                assert(lower <= upper && upper <= length, "slides slicing index out of bounds");
+
+                lower *= _stepSize;
+                upper *= _stepSize;
+
+                immutable len = _source.length;
+                // notice that we only need to move for windowSize - 1 to the right
+                // [0, 1, 2, 3] - slides(2) -> [[0, 1], [1, 2], [2, 3]]
+                // [0, 1, 2, 3] - slides(3) -> [[0, 1, 2], [1, 2, 3]]
+                return typeof(this)(_source[min(lower, len) .. min(upper + _windowSize - 1, len)], _windowSize, _stepSize);
+            }
+        }
+        else static if (hasSliceToEnd)
+        {
+            //For slicing an infinite chunk, we need to slice the source to the end.
+            auto opSlice(size_t lower, size_t upper)
+            {
+                assert(lower <= upper, "slides slicing index out of bounds");
+
+                return typeof(this)(_source[lower * _stepSize .. $], _windowSize, _stepSize).takeExactly(upper - lower);
+            }
+        }
+
+        static if (isInfinite!Source)
+        {
+            static if (hasSliceToEnd)
+            {
+                private static struct DollarToken{}
+                DollarToken opDollar()
+                {
+                    return DollarToken();
+                }
+                //Slice to dollar
+                typeof(this) opSlice(size_t lower, DollarToken)
+                {
+                    lower *= _stepSize;
+                    return typeof(this)(_source[lower .. $], _windowSize, _stepSize);
+                }
+            }
+        }
+        else
+        {
+            alias ThisType = typeof(this);
+
+            //Dollar token carries a static type, with no extra information.
+            //It can lazily transform into _source.length on algorithmic
+            //operations such as : slides[$/2, $-1];
+            private static struct DollarToken
+            {
+                ThisType* mom;
```

Comment:
no need to save the pointer, just the length (can't change for the lifetime of the token)